File: LeetCode/39组合总和.cpp
class Solution {
public:
    void backtracking(vector<int> &candidates, int target, int ind) {
        if (ind == candidates.size()) return ;
        if (target == 0) {
            ans.emplace_back(temp);
            return ;
        }
        backtracking(candidates, target, ind + 1);
        if (target - candidates[ind] >= 0) {
            temp.emplace_back(candidates[ind]);
            backtracking(candidates, target - candidates[ind], ind);
            temp.pop_back();
        }
        return ;
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        backtracking(candidates, target, 0);
        return ans;
    }

private :
    vector<vector<int>> ans;
    vector<int> temp;
};

File: LeetCode/39组合总和.go
var ans [][]int
var num []int

func combinationSum(candidates []int, target int) [][]int {
	ans = [][]int{}
	num = []int{}
	backtracking(0, 0, candidates, target)
	return ans
}

func backtracking(sum int, ind int, candidates []int, target int) {
	if sum > target {
		return
	}
	if sum == target {
		temp := make([]int, len(num))
		copy(temp, num)
		ans = append(ans, temp)
		return
	}
	for i := ind; i < len(candidates); i++ {
		sum += candidates[i]
		num = append(num, candidates[i])
		backtracking(sum, i, candidates, target)
		num = num[:len(num)-1]
		sum -= candidates[i]
	}
	return
}